### PR TITLE
Comprehend Moderation ver 0.1

### PR DESCRIPTION
**Pull request by iamwrick**
This PR streamlines the #9665 configuration object and logic to enable chain moderation via the Amazon Comprehend service, allowing you to detect and redact PII, Toxic, and Intent information in the LLM prompt or answer.
This implementation simplifies the configuration and types it to specific class/model.

### Usage sample

```python


from langchain_experimental.comprehend_moderation import (BaseModerationConfig, 
                                 ModerationIntentConfig, 
                                 ModerationPiiConfig, 
                                 ModerationToxicityConfig
)

pii_config = ModerationPiiConfig(
    labels=["SSN"],
    redact=True,
    mask_character="X"
)

toxicity_config = ModerationToxicityConfig(
    threshold=0.5
)

intent_config = ModerationIntentConfig(
    threshold=0.5
)

moderation_config = BaseModerationConfig(
    filters=[pii_config, toxicity_config, intent_config]
)

comp_moderation_with_config = AmazonComprehendModerationChain(
    moderation_config=moderation_config, #specify the configuration
    client=comprehend_client,            #optionally pass the Boto3 Client
    verbose=True
)

template = """Question: {question}

Answer:"""

prompt = PromptTemplate(template=template, input_variables=["question"])

responses = [
    "Final Answer: A credit card number looks like 1289-2321-1123-2387. A fake SSN number looks like 323-22-9980. John Doe's phone number is (999)253-9876.", 
    "Final Answer: This is a really shitty way of constructing a birdhouse. This is fucking insane to think that any birds would actually create their motherfucking nests here."
]
llm = FakeListLLM(responses=responses)

llm_chain = LLMChain(prompt=prompt, llm=llm)

chain = ( 
    prompt 
    | comp_moderation_with_config 
    | {llm_chain.input_keys[0]: lambda x: x['output'] }  
    | llm_chain 
    | { "input": lambda x: x['text'] } 
    | comp_moderation_with_config 
)

try:
    response = chain.invoke({"question": "A sample SSN number looks like this 123-456-7890. Can you give me some more samples?"})
except Exception as e:
    print(str(e))
else:
    print(response['output'])

```

### Output

```python

> Entering new AmazonComprehendModerationChain chain...
Running AmazonComprehendModerationChain...
Running pii Validation...
Running toxicity Validation...
Running intent Validation...

> Finished chain.


> Entering new AmazonComprehendModerationChain chain...
Running AmazonComprehendModerationChain...
Running pii Validation...
Running toxicity Validation...
Running intent Validation...

> Finished chain.
Final Answer: A credit card number looks like 1289-2321-1123-2387. A fake SSN number looks like XXXXXXXXXXXX John Doe's phone number is (999)253-9876.
```

### CODE REVIEW

File: libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py [hunk 1]
```diff
@@ -3,12 +3,13 @@
...
-    BaseModeration,
-)
+from langchain_experimental.comprehend_moderation.base_moderation import BaseModeration
 from langchain_experimental.comprehend_moderation.base_moderation_callbacks import (
     BaseModerationCallbackHandler,
 )
+from langchain_experimental.comprehend_moderation.base_moderation_config import (
+    BaseModerationConfig,
+)
 from langchain_experimental.pydantic_v1 import root_validator
```
Comment (3coins) [inline libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py:13]: This module is using Pydantic v1, but the Config models at the bottom are referencing v2. This won't work. Can we move everything to v1, as that's the default that LangChain is pulling in right now.

**Review by 3coins — CHANGES_REQUESTED** [review]

File: libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py (hunk 1 above)
Comment (nikhilkjha) [reply libs/experimental/langchain_experimental/comprehend_moderation/amazon_comprehend_moderation.py:13]: fixed.

**Review by 3coins — APPROVED** [review]
Looks good. Thanks for updating!

cc @baskaryan

**Review by hwchase17 — COMMENTED** [review]
i think all the examples that have
```
chain = ( 
    prompt 
    | comp_moderation_with_config 
    | {llm_chain.input_keys[0]: lambda x: x['output'] }  
    | llm_chain 
    | { "input": lambda x: x['text'] } 
    | comp_moderation_with_config 
)
```
should really read
```
llm = OpenAI()

chain = (
    prompt 
    | comprehend_moderation 
    | {"input": (lambda x: x['output'] ) | llm}
    | comprehend_moderation 
)
```
eg the LLMchain has prompt -> llm, but you want to stick content moderation in between the two, as far as I understand

**Comment by anjanvb** [comment]
> {"input": (lambda x: x['output'] ) | llm}

Copy that, yes that works as well. I tested out our first two example with your suggestion and it works. Will test out the other examples and change them.

**Comment by anjanvb** [comment]
New usage

```python
llm = FakeListLLM(responses=responses)

chain = ( 
    prompt 
    | comp_moderation_with_config 
    | { "input": (lambda x: x['output']) | llm }  
    | comp_moderation_with_config 
)

try:
    response = chain.invoke({"question": "A sample SSN number looks like this 123-456-7890. Can you give me some more samples?"})
except Exception as e:
    print(str(e))
else:
    print(response['output'])
```

**Comment by iamwrick** [comment]
@hwchase17 -  The #10104 looks like is merged but the PR status shows as open. The latest sample notebook is updated already by @anjanvb . Can you please pull the latest Notebook from the docs and let us know if we should close this PR.

**Comment by 3coins** [comment]
@iamwrick 
The PR is not merged yet. If you have all the updates in the PR, @hwchase17 should merge soon.

**Comment by anjanvb** [comment]
@3coins it does look like it was merged via #10153?

**Comment by 3coins** [comment]
> @3coins it does look like it was merged via #10153?

Yep, did not know there was another PR. If all updates are present in 10153, let's close this one.

**Comment by baskaryan** [comment]
> > @3coins it does look like it was merged via #10153?
> 
> Yep, did not know there was another PR. If all updates are present in 10153, let's close this one.

closing per ^

**Pull request closed without merging**